### PR TITLE
Add validation for namespace, metric and timestamp

### PR DESCRIPTION
**Pull request by markkuhn**
Related to: #131

* https://github.com/awslabs/aws-embedded-metrics-python/pull/91
* https://github.com/awslabs/aws-embedded-metrics-dotnet/pull/37
* https://github.com/awslabs/aws-embedded-metrics-java/pull/119

*Description of changes:*
Adding validation for metrics and namespace since validating dimensions alone doesn't make sense.

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

### CODE REVIEW

File: src/utils/Validator.ts [hunk 1]
```diff
@@ -70,4 +74,107 @@ export class Validator {
...
+      throw new InvalidMetricError(`Metric key ${key} must include at least one non-whitespace character`);
+    }
+
+    if (key.length > Constants.MAX_METRIC_NAME_LENGTH) {
+      throw new InvalidMetricError(
+        `Metric key ${key} must not exceed maximum length ${Constants.MAX_METRIC_NAME_LENGTH}`,
+      );
+    }
+
+    if (isNaN(value) || value === -Infinity || value === Infinity) {
```
Comment (evanuk) [inline src/utils/Validator.ts:98]: Can we use `Number.isNaN` here? `isNaN` has some weird behaviour: https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/isNaN#confusing_special-case_behavior

Comment (evanuk) [reply src/utils/Validator.ts:98]: Lets add some more `putMetric` validation tests as well to some of these examples of weird input/behaviour.

Comment (markkuhn) [reply src/utils/Validator.ts:98]: The only weird inputs I can give are `NaN` and `Infinity` given that the value must be of type `number`.

**Comment by markkuhn** [comment]
Update:

- The `faker` dependency currently used is deprecated/no longer maintained so I replaced it with `@faker-js/faker`.
- Updated TypeScript to 4.8.2
- Updated eslint to 8.23.0 and made some minor changes based on the new rules
- Since the min/max Metric values specified in the [specs](https://docs.aws.amazon.com/AmazonCloudWatch/latest/APIReference/API_MetricDatum.html) are outside the `Number` range, we will be validating against `Number.MAX_SAFE_INTEGER` and `-Number.MAX_SAFE_INTEGER`
- changed `isNaN()` to `Number.isNaN()`

[Interesting read ](https://www.bleepingcomputer.com/news/security/dev-corrupts-npm-libs-colors-and-faker-breaking-thousands-of-apps/)on what happened to faker and how the creator starting pushing malicious updates.

**Review by evanuk — APPROVED** [review]

**Review by Himtanaya — APPROVED** [review]

**Pull request merged**